src/libfsm/print/vmops.c

```diff
@@ -0,0 +1,404 @@
+
+	/* TODO: we'll need to heed cp for e.g. lx's codegen */
+	(void) cp;
+
+	if (!dfavm_compile_ir(&a, ir, vm_opts)) {
+		return -1;
+	}
+
+	for (op = a.linked; op != NULL; op = op->next) {
+		if (op->num_incoming > 0) {
+			print_label(f, op, opt, dialect);
+		}
+		switch (op->instr) {
+		case VM_OP_STOP:
+			print_cond(f, op, opt, dialect);
+			print_end(f, op, opt,  dialect, op->u.stop.end_bits, ir);
+			break;
+
+		case VM_OP_FETCH:
+			print_fetch(f, opt, dialect);
+			print_end(f, op, opt, dialect, op->u.fetch.end_bits, ir);
+			break;
+
+		case VM_OP_BRANCH:
+			print_cond(f, op, opt, dialect);
+			print_branch(f, op, dialect);
+			break;
+
+		default:
+			assert(!"unreached");
+			break;
+		}
+	}
+
+	dfavm_opasm_finalize_op(&a);
+
+	return 0;
+}
+
+static void
+fsm_print_vmops_complete(FILE *f, const struct ir *ir, const struct fsm_options *opt, enum vmops_dialect dialect)
+{
+	/* TODO: currently unused, but must be non-NULL */
+	const char *cp = "";
+
+	assert(f != NULL);
+	assert(ir != NULL);
+	assert(opt != NULL);
+
+	switch (dialect) {
+	case VMOPS_H:
+	case VMOPS_MAIN:
+		return;
+	default:
+		break;
+	}
+
+	(void) fsm_print_vmopsfrag(f, ir, opt, dialect, cp,
+		opt->leaf != NULL ? opt->leaf : leaf, opt->leaf_opaque);
+}
+
+void
+fsm_print_vmops(FILE *f, const struct fsm *fsm, enum vmops_dialect dialect)
+{
+	struct ir *ir;
+	const char *prefix;
+	const char *package_prefix;
+
+	assert(f != NULL);
+	assert(fsm != NULL);
+	assert(fsm->opt != NULL);
+
+	ir = make_ir(fsm);
+	if (ir == NULL) {
+		return;
+	}
+
+	/* henceforth, no function should be passed struct fsm *, only the ir and options */
+
+	if (fsm->opt->prefix != NULL) {
+		prefix = fsm->opt->prefix;
+	} else {
+		prefix = "fsm_";
+	}
+
+	if (fsm->opt->fragment) {
+		fsm_print_vmops_complete(f, ir, fsm->opt, dialect);
+		return;
+	}
+
+	if (fsm->opt->package_prefix != NULL) {
+		package_prefix = fsm->opt->package_prefix;
+	} else {
+		package_prefix = prefix;
+	}
+
+	switch (dialect) {
+	case VMOPS_C:
+		fprintf(f, "#include <stdint.h>\n\n");
+		fprintf(f, "#ifndef LIBFSM_VMOPS_H\n");
+		fprintf(f, "#include \"vmops.h\"\n");
+		fprintf(f, "#endif /* LIBFSM_VMOPS_H */\n");
+		fprintf(f, "struct op %sOps[] = {\n", prefix);
+		fsm_print_vmops_complete(f, ir, fsm->opt, dialect);
+		fprintf(f, "\t};\n");
+		break;
+
+	case VMOPS_H:
+		fprintf(f, "#ifndef LIBFSM_VMOPS_H\n");
+		fprintf(f, "#define LIBFSM_VMOPS_H\n");
+		fprintf(f, "#include <stdint.h>\n\n");
+		fprintf(f, "enum vmOp { opEOF, opLT, opLE, opEQ, opNE, opGE, opGT, opALWAYS};\n");
+		fprintf(f, "enum actionOp { actionRET, actionGOTO };\n");
+		fprintf(f, "struct op { enum vmOp op; char c; enum actionOp action; int32_t arg; };\n\n");
+		fprintf(f, "#endif /* LIBFSM_VMOPS_H */\n");
+		break;
+
+	case VMOPS_MAIN:
+		fprintf(f, "#include <stdio.h>\n");
+		fprintf(f, "#include <stdlib.h>\n\n");
+		fprintf(f, "#ifndef LIBFSM_VMOPS_H\n");
+		fprintf(f, "#include \"vmops.h\"\n");
+		fprintf(f, "#endif /* LIBFSM_VMOPS_H */\n");
```

Comment:
cheesy, but okay (and no less cheesy than the other C output)

Reply:
I would love a better solution here but couldn't come up with anything.